src/node/rpc/jwt_management.h

```diff
@@ -192,38 +243,132 @@ namespace ccf
-        if (issuer_ == issuer)
+    std::set<std::string> kids_with_new_constraints;
+    key_issuers->foreach([&existing_kids,
+                          &issuer_constraints,
+                          &kids_with_new_constraints,
+                          &issuer](const auto& kid, const auto& issuers) {
+      for (const auto& issuer_with_constraint : issuers)
+      {
+        if (issuer == issuer_with_constraint.issuer)
         {
           existing_kids.insert(kid);
+
+          const auto it = issuer_constraints.find(kid);
+          if (
+            it != issuer_constraints.end() &&
+            it->second == issuer_with_constraint.constraint)
+          {
+            kids_with_new_constraints.insert(kid);
+          }
+
+          break; // 1 issuer <-> 1 kid
         }
-        return true;
-      });
+      }
+      return true;
+    });
 
     for (auto& [kid, der] : new_keys)
     {
-      if (!existing_kids.contains(kid))
+      const bool new_kid = !existing_kids.contains(kid);
+      const bool new_constraint = !kids_with_new_constraints.contains(kid);
+
+      if (new_kid)
       {
         keys->put(kid, der);
-        key_issuer->put(kid, issuer);
+      }
+
+      if (new_kid || new_constraint)
+      {
+        JwtIssuerWithConstraint value{issuer, std::nullopt};
+        const auto it = issuer_constraints.find(kid);
+        if (it != issuer_constraints.end())
+        {
+          value.constraint = it->second;
+        }
+
+        LOG_INFO_FMT(
+          "Save JWT issuer for kid {} where issuer: {}, issuer constraint: {}",
+          kid,
+          value.issuer,
+          value.constraint);
+
+        auto existing_issuers = key_issuers->get(kid);
+        if (existing_issuers)
+        {
+          const auto prev = find_if(
+            existing_issuers->begin(),
+            existing_issuers->end(),
+            [&](const auto& issuer_with_constraint) {
+              return issuer_with_constraint.issuer == issuer;
+            });
+
+          if (prev != existing_issuers->end())
+          {
+            *prev = value;
+          }
+          else
+          {
+            existing_issuers->push_back(std::move(value));
+          }
+          key_issuers->put(kid, *existing_issuers);
+        }
+        else
+        {
+          key_issuers->put(kid, std::vector<JwtIssuerWithConstraint>{value});
+        }
       }
     }
 
     for (auto& kid : existing_kids)
     {
       if (!new_keys.contains(kid))
       {
-        keys->remove(kid);
-        key_issuer->remove(kid);
+        auto updated = key_issuers->get(kid);
+        updated->erase(
+          std::remove_if(
+            updated->begin(),
+            updated->end(),
+            [&](const auto& issuer_with_constraint) {
+              return issuer_with_constraint.issuer == issuer;
+            }),
+          updated->end());
+
+        if (updated->empty())
+        {
+          keys->remove(kid);
+          key_issuers->remove(kid);
+        }
```

Comment:
Non-critical, but maybe we should try to clean up the `Legacy` tables here as well? So don't bother trying to read from them to populate `existing_kids`, but if we're removing from the new tables then we could also try `legacy_keys->remove(kid)`?

Reply:
See https://github.com/microsoft/CCF/pull/6175/files/00972768b191b644e90e189700f8622480ad5c80#r1620429945